sinks/xray/xray.go

```diff
@@ -0,0 +1,220 @@
+package xray
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"hash/crc32"
+	"math"
+	"net"
+	"regexp"
+	"strconv"
+	"sync/atomic"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/stripe/veneur/protocol"
+	"github.com/stripe/veneur/sinks"
+	"github.com/stripe/veneur/ssf"
+	"github.com/stripe/veneur/trace"
+	"github.com/stripe/veneur/trace/metrics"
+)
+
+const subsegmentType = "subsegment"
+
+var segmentHeader = []byte(`{"format": "json", "version": 1}` + "\n")
+
+const XRayTagNameClientIP = "xray_client_ip"
+
+type XRaySegmentHTTPRequest struct {
+	Method        string `json:"method,omitempty"`
+	ClientIP      string `json:"client_ip,omitempty"`
+	URL           string `json:"url,omitempty"`
+	UserAgent     string `json:"user_agent,omitempty"`
+	XForwardedFor string `json:"x_forwarded_for,omitempty"`
+}
+
+type XRaySegmentHTTPResponse struct {
+	Status int `json:"status,omitempty"`
+}
+
+type XRaySegmentHTTP struct {
+	Request  XRaySegmentHTTPRequest  `json:"request,omitempty"`
+	Response XRaySegmentHTTPResponse `json:"response,omitempty"`
+}
+
+// XRaySegment is a trace segment for X-Ray as defined by:
+// https://docs.aws.amazon.com/xray/latest/devguide/xray-api-segmentdocuments.html
+type XRaySegment struct {
+	Name        string            `json:"name"`
+	ID          string            `json:"id"`
+	TraceID     string            `json:"trace_id"`
+	ParentID    string            `json:"parent_id,omitempty"`
+	StartTime   float64           `json:"start_time"`
+	EndTime     float64           `json:"end_time"`
+	SegmentType string            `json:"type,omitempty"`
+	Namespace   string            `json:"namespace"`
+	Error       bool              `json:"error"`
+	Annotations map[string]string `json:"annotations,omitempty"`
+	Metadata    map[string]string `json:"metadata,omitempty"`
+	HTTP        XRaySegmentHTTP   `json:"http,omitempty"`
+}
+
+// XRaySpanSink is a sink for spans to be sent to AWS X-Ray.
+type XRaySpanSink struct {
+	daemonAddr      string
+	traceClient     *trace.Client
+	conn            *net.UDPConn
+	sampleThreshold uint32
+	commonTags      map[string]string
+	log             *logrus.Logger
+	spansDropped    int64
+	spansHandled    int64
+	nameRegex       *regexp.Regexp
+}
+
+var _ sinks.SpanSink = &XRaySpanSink{}
+
+// NewXRaySpanSink creates a new instance of a XRaySpanSink.
+func NewXRaySpanSink(daemonAddr string, sampleRatePercentage int, commonTags map[string]string, log *logrus.Logger) (*XRaySpanSink, error) {
+
+	log.WithFields(logrus.Fields{
+		"Address": daemonAddr,
+	}).Info("Creating X-Ray client")
+
+	var sampleThreshold uint32
+	if sampleRatePercentage <= 0 || sampleRatePercentage > 100 {
+		return nil, errors.New("Span sample rate percentage must be greater than 0% and less than or equal to 100%")
```

Comment:
I'd rather have the zero value be a safe default, so that specifying the config option is optional.